cluster_pipeline.go

```diff
@@ -0,0 +1,134 @@
+package redis
+
+// ClusterPipeline is not thread-safe.
+type ClusterPipeline struct {
+	commandable
+
+	cmds    []Cmder
+	cluster *ClusterClient
+	closed  bool
+}
+
+// Pipeline creates a new pipeline which is able to execute commands
+// against multiple shards.
+func (c *ClusterClient) Pipeline() *ClusterPipeline {
+	pipe := &ClusterPipeline{
+		cluster: c,
+		cmds:    make([]Cmder, 0, 10),
+	}
+	pipe.commandable.process = pipe.process
+	return pipe
+}
+
+func (c *ClusterPipeline) process(cmd Cmder) {
+	c.cmds = append(c.cmds, cmd)
+}
+
+// Close marks the pipeline as closed
+func (c *ClusterPipeline) Close() error {
+	c.closed = true
+	return nil
+}
+
+// Discard resets the pipeline and discards queued commands
+func (c *ClusterPipeline) Discard() error {
+	if c.closed {
+		return errClosed
+	}
+	c.cmds = c.cmds[:0]
+	return nil
+}
+
+func (c *ClusterPipeline) Exec() (cmds []Cmder, retErr error) {
+	if c.closed {
+		return nil, errClosed
+	}
+
+	cmds = c.cmds
+	c.cmds = make([]Cmder, 0, 10)
+	if len(cmds) == 0 {
+		return []Cmder{}, nil
+	}
+
+	cmdsMap := make(map[string][]Cmder)
+	for _, cmd := range cmds {
+		slot := hashSlot(cmd.clusterKey())
+		addrs := c.cluster.slotAddrs(slot)
+
+		var addr string
+		if len(addrs) > 0 {
+			addr = addrs[0] // First address is master.
+		}
+
+		cmdsMap[addr] = append(cmdsMap[addr], cmd)
+	}
+
+	for attempt := 0; attempt < c.cluster.opt.getMaxRedirects(); attempt++ {
+		failedCmds := make(map[string][]Cmder)
+
+		for addr, cmds := range cmdsMap {
+			var client *Client
+			if addr != "" {
+				client = c.cluster.getClient(addr)
```

Comment:
Or at least move that block into a helper method. We are using a very similar logic in `ClusterClient.process`